### PR TITLE
DSM Radio Revisions

### PR DESCRIPTION
**Pull request by NotFastEnuf**
Autodetect DSM protocol based on serial ID byte and handle accordingly.

### CODE REVIEW

File: src/main/rx/rx_unified_dsm.c [hunk 1]
```diff
@@ -37,26 +36,52 @@ extern uint8_t ready_for_next_telemetry;
 #define USART usart_port_defs[serial_rx_port]
 #define SPECTRUM_BIND_PIN usart_port_defs[profile.serial.rx].rx_pin
 
-void rx_serial_process_dsmx() {
+static uint8_t dsm_protocol = 0;
+#define DSMX_11_2048 0xb2
+#define DSMX_22_2048 0xa2
+#define DSM2_11_2048 0x12
+#define DSM2_22_1024 0x01
```
Comment (bkleiner) [inline src/main/rx/rx_unified_dsm.c:43]: lets convert these to an enum and move that to the top of the file, after the includes.
`dms_protocol` should use this enum type.

File: src/main/rx/rx_unified_dsm.c [hunk 2]
```diff
@@ -37,26 +36,52 @@ extern uint8_t ready_for_next_telemetry;
...
+    dsm_offset = 1024.0f;
+  }
+  if (dsm_protocol == DSM2_22_1024){
+    // 10 bit frames
+    spek_chan_shift = 2;
+    spek_chan_mask = 0x03;
+    dsm_channel_count = 7;
+    dsm_scalefactor = (0.29354210f / DSM_SCALE_PERCENT);
+    dsm_offset = 512.0f;
+  }
```
Comment (bkleiner) [inline src/main/rx/rx_unified_dsm.c:84]: lets use a switch statement here. bet advised that you switch statement can handle multiple cases the same way:
```c
switch (dsm_protocol) {
case DSMX_11_2048:
case DSMX_22_2048:
     ...code for 11bit frames
     break;
}
```

File: src/main/rx/rx_unified_dsm.c [hunk 3]
```diff
@@ -37,26 +36,52 @@ extern uint8_t ready_for_next_telemetry;
...
+    dsm_scalefactor = (0.14662756f / DSM_SCALE_PERCENT);
+    dsm_offset = 1024.0f;
+  }
+  if (dsm_protocol == DSM2_11_2048){
+    // 10 bit frames
+    spek_chan_shift = 2;
+    spek_chan_mask = 0x03;
+    dsm_channel_count = 7;
+    dsm_scalefactor = (0.14662756f / DSM_SCALE_PERCENT);
+    dsm_offset = 1024.0f;
```
Comment (bkleiner) [inline src/main/rx/rx_unified_dsm.c:75]: the [spektrum datasheet](https://www.spektrumrc.com/ProdInfo/Files/Remote%20Receiver%20Interfacing%20Rev%20A.pdf) (Section 8) says DSM2 2048 also uses 11bit channels.
did you verify this?

**Review by bkleiner — CHANGES_REQUESTED** [review]

File: src/main/rx/rx_unified_dsm.c (hunk 3 above)
Comment (NotFastEnuf) [reply src/main/rx/rx_unified_dsm.c:75]: No, this configuration is unverified.   I am unable to test this particular configuration with any gear I have available.  The above combination was my best guess.

Comment (bkleiner) [reply src/main/rx/rx_unified_dsm.c:75]: all references i have found point towards it being 11bit as well, so lets go with that.

File: src/main/rx/rx_unified_dsm.c [hunk 4]
```diff
@@ -13,6 +13,15 @@
 #include "profile.h"
 #include "util.h"
 
+typedef enum {
+  PROTOCOL_INVALID,
+  DSMX_11_2048,
+  DSMX_22_2048,
+  DSM2_11_2048,
+  DSM2_22_1024,
```
Comment (bkleiner) [inline src/main/rx/rx_unified_dsm.c:21]: you can have these equal whatever you want, eg:
`DSM2_22_1024 = 0x01,`

File: src/main/rx/rx_unified_dsm.c [hunk 5]
```diff
@@ -36,20 +45,22 @@ extern uint8_t ready_for_next_telemetry;
...
-#define DSMX_22_2048 0xa2
-#define DSM2_11_2048 0x12
-#define DSM2_22_1024 0x01
+uint8_t check_dsm_id (uint8_t id_byte){
+  if (id_byte == 0xb2) return DSMX_11_2048;
+  if (id_byte == 0xa2) return DSMX_22_2048;
+  if (id_byte == 0x12) return DSM2_11_2048;
+  if (id_byte == 0x01) return DSM2_22_1024;
+  return PROTOCOL_INVALID;
+}
```
Comment (bkleiner) [inline src/main/rx/rx_unified_dsm.c:54]: lets have the enum entries be the id values, that will remove the need for this function

File: src/main/rx/rx_unified_dsm.c [hunk 6]
```diff
@@ -13,6 +13,15 @@
...
 #include "util.h"
 
+typedef enum {
+  PROTOCOL_INVALID,
+  DSMX_11_2048,
+  DSMX_22_2048,
+  DSM2_11_2048,
+  DSM2_22_1024,
+} dsm_protocol_t;
+dsm_protocol_t dsm_protocol;
```
Comment (bkleiner) [inline src/main/rx/rx_unified_dsm.c:23]: `static` and a init value please

**Review by bkleiner — CHANGES_REQUESTED** [review]

File: src/main/rx/rx_unified_dsm.c (hunk 6 above)
Comment (NotFastEnuf) [reply src/main/rx/rx_unified_dsm.c]: ..Thank you

**Review by bkleiner — APPROVED** [review]
lovely, did you already try it with the config? as far as i can tell it should just work.

**Comment by NotFastEnuf** [comment]
Everything from the perspective of the config remains the same and tests ok there.  All dsm variants are generalized as one DSM protocol.

**Pull request merged**